src/orchestrator.ts

```diff
@@ -243,31 +243,80 @@ export class Orchestrator {
         : Task {
         const newAction = new CallActivityWithRetryAction(name, retryOptions, input);
 
-        for (let attempt = 1; attempt <= retryOptions.maxNumberOfAttempts; attempt++) {
-            const taskScheduled = this.findTaskScheduled(state, name);
-            const taskCompleted = this.findTaskCompleted(state, taskScheduled);
-            const taskFailed = this.findTaskFailed(state, taskScheduled);
-            const taskRetryTimer = this.findRetryTimer(state, taskFailed);
-            const taskRetryTimerFired = this.findTimerFired(state, taskRetryTimer);
-            this.setProcessed([ taskScheduled, taskCompleted, taskFailed, taskRetryTimer, taskRetryTimerFired ]);
-
-            if (!taskScheduled) { break; }
-
-            if (taskCompleted) {
-                const result = this.parseHistoryEvent(taskCompleted);
-
-                return TaskFactory.SuccessfulTask(newAction, result, taskCompleted.Timestamp, taskCompleted.TaskScheduledId, state.indexOf(taskCompleted));
-            } else if (taskFailed
-                && taskRetryTimer
-                && attempt >= retryOptions.maxNumberOfAttempts) {
-                return TaskFactory.FailedTask(
-                    newAction,
-                    taskFailed.Reason,
-                    taskFailed.Timestamp,
-                    taskFailed.TaskScheduledId,
-                    state.indexOf(taskFailed),
-                    new DurableError(taskFailed.Reason),
-                );
+        let attempt = 1;
+        let taskScheduled: TaskScheduledEvent | undefined;
+        let taskFailed: TaskFailedEvent | undefined;
+        let taskRetryTimer: TimerCreatedEvent | undefined;
+        for (let i = 0; i < state.length; i++) {
```

Comment:
nit: can we do a `foreach` / `for ... in`-style loop here?

Reply:
I guess in TS its a `for...of`

Reply:
I think this was probably a case of me trying to micro-optimize the perf of this loop. Probably not worth it.

Reply:
I'm leaning away from this. I went with the explicit for-loop as we use the index in the loop body.